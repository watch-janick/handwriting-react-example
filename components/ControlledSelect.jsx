import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import autobind from 'autobind-decorator'

import {
  Input,
  InputGroup,
  InputGroupAddon,
} from 'reactstrap'

/**
 * Component used to display a selectable list of handwritings
 *
 * @extends PureComponent
 */
class ControlledSelect extends PureComponent {
  static propTypes = {
    selectChoice: PropTypes.func.isRequired,
    choices: PropTypes.arrayOf(PropTypes.shape),
    selectedChoice: PropTypes.string,
  }

  @autobind // use autobind to make this available in textChanged
  onChange(e) {
    const { selectChoice } = this.props
    const choiceId = e.target.value

    selectChoice(choiceId)
  }

  render() {
    const {
      choices,
      selectedChoice,
    } = this.props

    const value = selectedChoice || ''

    return (
      <InputGroup>
        <InputGroupAddon addonType="prepend">Handwriting style</InputGroupAddon>
        <Input
          type="select"
          onChange={this.onChange}
          value={value}
          id="handwriting-styles"
        >
          {choices && choices.map(({ id, title }) =>
            <option key={id} value={id}>{title}</option>
          )}
        </Input>
      </InputGroup>
    )
  }
}

export default ControlledSelect