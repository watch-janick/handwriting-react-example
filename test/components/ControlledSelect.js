import ControlledSelect from '../../components/ControlledSelect'

describe('ControlledSelect', () => {
  it('should render correctly with minimal attributes', () => {
    const attrs = {
      selectChoice: fn,
    }
    const component = shallow(<ControlledSelect {...attrs} />)

    expect(component).to.matchSnapshot()
  })

  it('should render correctly with choices', () => {
    const attrs = {
      selectChoice: fn,
      choices: [{
        id: '31SB3NWR00E0',
        title: 'Alder',
      }, {
        id: '31SF81NG00ES',
        title: 'Kelly',
      }],
    }
    const component = shallow(<ControlledSelect {...attrs} />)

    expect(component).to.matchSnapshot()
  })

  it('should correctly display selected choice', () => {
    const selectedChoice = '31SF81NG00ES'
    const attrs = {
      selectChoice: fn,
      choices: [{
        id: '31SB3NWR00E0',
        title: 'Alder',
      }, {
        id: selectedChoice,
        title: 'Kelly',
      }],
      selectedChoice,
    }
    const component = shallow(<ControlledSelect {...attrs} />)
    const select = component.find('#handwriting-styles')

    expect(select.props().value).to.equal(selectedChoice)
  })

  it('should call selectChoice when value changes', () => {
    const selectChoice = spy()
    const value = '31SF81NG00ES'
    const attrs = {
      selectChoice,
      choices: [{
        id: '31SB3NWR00E0',
        title: 'Alder',
      }, {
        id: value,
        title: 'Kelly',
      }],
    }
    const component = shallow(<ControlledSelect {...attrs} />)
    const select = component.find('#handwriting-styles')

    select.simulate('change', { target : { value } } )

    expect(selectChoice.calledOnceWith(value)).to.equal(true)
  })
})