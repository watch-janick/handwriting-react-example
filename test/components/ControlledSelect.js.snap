// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlledSelect should render correctly with choices 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ControlledSelect
    choices={
      Array [
        Object {
          "id": "31SB3NWR00E0",
          "title": "Alder",
        },
        Object {
          "id": "31SF81NG00ES",
          "title": "Kelly",
        },
      ]
    }
    selectChoice={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <InputGroupAddon
          addonType="prepend"
          tag="div"
        >
          Handwriting style
        </InputGroupAddon>,
        <Input
          id="handwriting-styles"
          onChange={[Function]}
          type="select"
          value=""
        >
          <option
            value="31SB3NWR00E0"
          >
            Alder
          </option>
          <option
            value="31SF81NG00ES"
          >
            Kelly
          </option>
        </Input>,
      ],
      "tag": "div",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "addonType": "prepend",
          "children": "Handwriting style",
          "tag": "div",
        },
        "ref": null,
        "rendered": "Handwriting style",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <option
              value="31SB3NWR00E0"
            >
              Alder
            </option>,
            <option
              value="31SF81NG00ES"
            >
              Kelly
            </option>,
          ],
          "id": "handwriting-styles",
          "onChange": [Function],
          "type": "select",
          "value": "",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "31SB3NWR00E0",
            "nodeType": "host",
            "props": Object {
              "children": "Alder",
              "value": "31SB3NWR00E0",
            },
            "ref": null,
            "rendered": "Alder",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "31SF81NG00ES",
            "nodeType": "host",
            "props": Object {
              "children": "Kelly",
              "value": "31SF81NG00ES",
            },
            "ref": null,
            "rendered": "Kelly",
            "type": "option",
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <InputGroupAddon
            addonType="prepend"
            tag="div"
          >
            Handwriting style
          </InputGroupAddon>,
          <Input
            id="handwriting-styles"
            onChange={[Function]}
            type="select"
            value=""
          >
            <option
              value="31SB3NWR00E0"
            >
              Alder
            </option>
            <option
              value="31SF81NG00ES"
            >
              Kelly
            </option>
          </Input>,
        ],
        "tag": "div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "addonType": "prepend",
            "children": "Handwriting style",
            "tag": "div",
          },
          "ref": null,
          "rendered": "Handwriting style",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <option
                value="31SB3NWR00E0"
              >
                Alder
              </option>,
              <option
                value="31SF81NG00ES"
              >
                Kelly
              </option>,
            ],
            "id": "handwriting-styles",
            "onChange": [Function],
            "type": "select",
            "value": "",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "31SB3NWR00E0",
              "nodeType": "host",
              "props": Object {
                "children": "Alder",
                "value": "31SB3NWR00E0",
              },
              "ref": null,
              "rendered": "Alder",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "31SF81NG00ES",
              "nodeType": "host",
              "props": Object {
                "children": "Kelly",
                "value": "31SF81NG00ES",
              },
              "ref": null,
              "rendered": "Kelly",
              "type": "option",
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`ControlledSelect should render correctly with minimal attributes 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ControlledSelect
    selectChoice={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <InputGroupAddon
          addonType="prepend"
          tag="div"
        >
          Handwriting style
        </InputGroupAddon>,
        <Input
          id="handwriting-styles"
          onChange={[Function]}
          type="select"
          value=""
        />,
      ],
      "tag": "div",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "addonType": "prepend",
          "children": "Handwriting style",
          "tag": "div",
        },
        "ref": null,
        "rendered": "Handwriting style",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": undefined,
          "id": "handwriting-styles",
          "onChange": [Function],
          "type": "select",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <InputGroupAddon
            addonType="prepend"
            tag="div"
          >
            Handwriting style
          </InputGroupAddon>,
          <Input
            id="handwriting-styles"
            onChange={[Function]}
            type="select"
            value=""
          />,
        ],
        "tag": "div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "addonType": "prepend",
            "children": "Handwriting style",
            "tag": "div",
          },
          "ref": null,
          "rendered": "Handwriting style",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": undefined,
            "id": "handwriting-styles",
            "onChange": [Function],
            "type": "select",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
